### PR TITLE
fix codegen bugs with llvm-gc-root pass

### PR DESCRIPTION
**Pull request by vtjnash**
ensures the return type of literal_pointer_val is always just a simple pointer, and always of the same type (T_pjlvalue)

### CODE REVIEW

File: src/codegen.cpp [hunk 1]
```diff
@@ -3726,10 +3726,22 @@ static jl_cgval_t emit_expr(jl_codectx_t &ctx, jl_value_t *expr)
                 mod = jl_globalref_mod(mn);
                 mn = (jl_value_t*)jl_globalref_name(mn);
             }
-            if (jl_symbol_name((jl_sym_t*)mn)[0] == '@')
-                jl_errorf("macro definition not allowed inside a local scope");
-            name = literal_pointer_val(ctx, mn);
-            bnd = jl_get_binding_for_method_def(mod, (jl_sym_t*)mn);
+            JL_TRY {
```
Comment (Keno) [inline src/codegen.cpp]: Unrelated change?

**Review by Keno — APPROVED** [review]
Looks semantically valid to me. If this fixes an identified bug, would be good to have a test to that extent.

File: src/codegen.cpp (hunk 1 above)
Comment (vtjnash) [inline src/codegen.cpp]:
```
julia> begin
         for i = 1:1 end
         macro g() end
       end
ERROR: error compiling anonymous: macro definition not allowed inside a local scope
```
vs.
```
julia> begin
         for i = 1:1 end
         macro g() end
       end
ERROR: macro definition not allowed inside a local scope
Stacktrace:
 [1] macro expansion at ./REPL[1]:3 [inlined]
 [2] anonymous at ./<missing>:?
```

Both results are wrong, but at least it now usually won't corrupt the C++ stack.

**Comment by tkelman** [comment]
did this get put into #22984 accidentally? agreed on a test if possible, how'd you find this or what is it fixing?

**Comment by vtjnash** [comment]
It wasn't accidental - the case that triggered my review and fixes here were from writing tests for that PR.
